### DISCUSSION TITLE
How are you applying effects like PitchBend to the tracks?

### DISCUSSION THREAD

**Discussion by corysimmons** [ryohey/signal#317] [Q&A] [answered]
I'm just curious what package or approach you're using to apply your effects. I tried to reverse engineer a bit but figured it'd be easier to just ask. :)

**Comment by ryohey** [comment, marked as answer]
Hi @corysimmons, The pitch is changed by changing the playback speed of the sample. The synthesizer code is implemented in the [wavelet](https://github.com/ryohey/wavelet) package. [This code](https://github.com/ryohey/wavelet/blob/main/lib/src/processor/WavetableOscillator.ts#L79) changes the playback speed of the sample according to the pitch bend value. The samples are PCM data, that is, an array of amplitudes. If you advance the index of the array one at a time, the pitch is the same as the PCM data, but if you advance the index two at a time, for example, the pitch is an octave higher. This means that the pitch bend value is converted to the size of the index advancement.

**Reply by corysimmons** [reply]
Ah, thank you! I'll look at this. Is there any reason you decided against using something like Tone.js?

**Reply by ryohey** [reply]
A built-in GM sound source was required for first-class MIDI support. I also needed high performance to be able to play a very large number of notes, so I created my own implementation using the new Audio Worklet.

**Reply by corysimmons** [reply]
Did you happen to run benchmarks on Tone? I am making a DAW and haven't ran into problems with Tone yet but maybe I need to stress test it.

Anyway, thank you for the help! Very cool project! ❤️
